### ISSUE TITLE
uninit condition in PixelTrackProducerFromSoAPhase1

### ISSUE THREAD

**Issue by slava77** [cms-sw/cmssw#42374]
While investigating https://its.cern.ch/jira/browse/CMSHLT-2881
I ran valgrind (in a single thread version)
`valgrind --tool=memcheck --suppressions=$CMSSW_RELEASE_BASE/src/Utilities/ReleaseScripts/data/cms-valgrind-memcheck.supp --num-callers=20 --xml=yes --xml-file=valgrind-%p.xml --malloc-fill=0xa5 --free-fill=0x5a cmsRunGlibC hlt_threads4_onlyPFJet140_0.py`

After ignoring seemingly unrelated uninit condition cases in ROOT, what stands out (a dozen or more) cases in 
https://github.com/cms-sw/cmssw/blob/1e813470368a11cefce5fcb8fabfe7c0db0f636f/RecoTracker/PixelTrackFitting/plugins/PixelTrackProducerFromSoA.cc#L178-L180

It looks like the upstream `CAHitNtupletGeneratorOnGPU::makeTuples` is producing some uninitialized content.

@AdrianoDee @vjmastra @mmusich

**Comment by slava77** [comment]
@missirol

**Comment by slava77** [comment]
I added printouts for pt values and in the first event it's 
```
Prepare to sort over 331
0 0.896421
1 0.75331
2 0.521295
3 1.82869
4 -2.87352e-16
5 0.540607
6 31.8181
7 -2.87352e-16
8 0.600337
```

where `-2.87352e-16` is coming from `--malloc-fill=0xa5`, the uninitialized values.

Considering how frequent these are, do I understand correctly that the uninitialized values are actually intentional on the creation side in the algorithm `makeTuples`?
And I guess the `PixelTrackProducerFromSoA` better not do the sort before thinning out the collection by quality.

**Comment by mmusich** [comment]
assign reconstruction
   * @silviodonato FYI, as sorting was originally introduced in https://github.com/cms-sw/cmssw/commit/62af9d468761541c14481ebbc9b34b64dd5b5195

**Comment by silviodonato** [comment]
Thank you @slava77  for the check. Looking at [this discussion](https://github.com/cms-sw/cmssw/pull/38065/files#r881565771) I see that only tracks with bad quality have an unassigned pT. These tracks are discarded later on ( https://github.com/cms-sw/cmssw/blob/master/RecoTracker/PixelTrackFitting/plugins/PixelTrackProducerFromSoA.cc#L189-L190 ).

In other words, the final "tracks" collection should not contain any track with unassigned pT  
https://github.com/cms-sw/cmssw/blob/master/RecoTracker/PixelTrackFitting/plugins/PixelTrackProducerFromSoA.cc#L251 .

**Comment by slava77** [comment]
> In other words, the final "tracks" collection should not contain any track with unassigned pT

OK.
Still, the valgrind error better be fixed even if it's benign here.

**Comment by Dr15Jones** [comment]
Also sorting with an uninitialized value is a dangerous proposition as the bit pattern for the value could happen to be a NaN. Sorting with a NaN leads to serious problems (usually a crash).

**Comment by silviodonato** [comment]
fix in https://github.com/cms-sw/cmssw/pull/42428

**Comment by mmusich** [comment]
Is this issue still active now that https://github.com/cms-sw/cmssw/pull/42428 is merged?

**Comment by slava77** [comment]
I'm closing this since it was addressed in #42428
